### PR TITLE
HS-789: Add node tag manangement: add and remove a tag from node(s)

### PR DESCRIPTION
**Pull request by thomasbigger584**
## Description
HS-789: Add node tag manangement: add and remove a tag from node(s)

Tag has a many to many relationship with Node with a degree of idempotency on create/remove

Creating a Tag will associate it with a Node, if its not already associated.
Removing a Tag will remove the Tag association with the node, and if there is no other associations it will delete the Tag.

Deleting a node will delete the associations of the tag, if there is no other associations it will delete the tag also.

Endpoints add/remove a list of tags.

## Jira link(s)
- https://issues.opennms.org/browse/HS-789

## Checklist
* [x] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [x] Appropriate reviewer(s) have been selected.
* [x] Jira issue(s) have been updated to "In Review".
* [x] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [x] Documentation has been updated as necessary.

### CODE REVIEW

File: inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Query("SELECT tag " +
+        "FROM Tag tag " +
+        "JOIN tag.nodes node " +
+        "WHERE tag.tenantId = :tenantId " +
+        "AND node.id = :nodeId " +
+        "AND tag.name = :name")
+    Optional<Tag> find(@Param("tenantId") String tenantId,
+                       @Param("nodeId") Long nodeId,
+                       @Param("name") String name);
```
Comment (yangli2014) [inline inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java:57]: We might need to find all tags by nodeId and tenantId

File: inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+@Repository
+public interface TagRepository extends JpaRepository<Tag, Long> {
+    Optional<Tag> findByTenantIdAndId(String tenantId, Long id);
+
+    @Query("SELECT tag " +
+        "FROM Tag tag " +
+        "WHERE tag.tenantId = :tenantId " +
+        "AND tag.name = :name")
+    Optional<Tag> find(@Param("tenantId") String tenantId,
```
Comment (yangli2014) [inline inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java:46]: It is not a big issue, but I prefer using JPA repo standard syntax findByxxxAndxxx to make it simple.

File: inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java (hunk 1 above)
Comment (thomasbigger584) [reply inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java:57]: yeah this is not done as its not related to this story

File: inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java (hunk 2 above)
Comment (thomasbigger584) [reply inventory/src/main/java/org/opennms/horizon/inventory/repository/TagRepository.java:46]: will update

**Review by yangli2014 — COMMENTED** [review]
Also, I think there might be a use case the user wants to add more than one tag to a node. So the user has to add one by one, it is better if we can let user add more want with some syntax such as the semicolon-separated string.

**Review by cgorantla — COMMENTED** [review]
Looks pretty good to me.

It aligns with our current thinking that tags are very generic.

As @yangli2014 said, we may want to add/delete multiple tags at the same time. 
May be it would be good to update services to accept lists.

**Comment by thomasbigger584** [comment]
> Looks pretty good to me.
> 
> It aligns with our current thinking that tags are very generic.
> 
> As @yangli2014 said, we may want to add/delete multiple tags at the same time. May be it would be good to update services to accept lists.

sure will make it a list instead

File: inventory/src/main/java/org/opennms/horizon/inventory/model/Tag.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    @NotNull
+    @Column(name = "name")
+    private String name;
+
+    @ManyToMany
+    @JoinTable(
+        name = "node_tag",
+        joinColumns = @JoinColumn(name = "tag_id"),
+        inverseJoinColumns = @JoinColumn(name = "node_id"))
+    List<Node> nodes = new ArrayList<>();
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/model/Tag.java:70]: nit :
```suggestion
    private List<Node> nodes = new ArrayList<>();
```

File: inventory/src/main/java/org/opennms/horizon/inventory/service/TagService.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Transactional
+    public List<TagDTO> addTags(String tenantId, TagCreateListDTO request) {
+        if (request.getTagsList().isEmpty()) {
+            return Collections.emptyList();
+        }
+        Node node = getNode(tenantId, request.getNodeId());
+
+        return request.getTagsList().stream()
+            .map(tagCreateDTO -> createSingle(tenantId, node, tagCreateDTO))
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/service/TagService.java:65]:
```suggestion
            .map(tagCreateDTO -> createTag(tenantId, node, tagCreateDTO))
```

File: inventory/src/main/java/org/opennms/horizon/inventory/service/TagService.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Transactional
+    public void removeTags(String tenantId, TagRemoveListDTO request) {
+        Node node = getNode(tenantId, request.getNodeId());
+
+        request.getTagIdsList().stream()
+            .map(Int64Value::getValue)
+            .map(tagId -> getTag(tenantId, tagId))
+            .toList()
+            .forEach(tag -> removeSingle(node, tag));
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/service/TagService.java:77]:
```suggestion
            .forEach(tag -> removeTagFromNode(node, tag));
```

**Review by cgorantla — COMMENTED** [review]
LGTM!  few minor comments

**Comment by yangli2014** [comment]
We also need to list all tags for giving node id to avoid unnecessary network traffic.

File: inventory/src/main/java/org/opennms/horizon/inventory/model/Tag.java (hunk 3 above)
Comment (thomasbigger584) [reply inventory/src/main/java/org/opennms/horizon/inventory/model/Tag.java:70]: good catch

**Comment by thomasbigger584** [comment]
> We also need to list all tags for giving node id to avoid unnecessary network traffic.

sure, can do that in another story

**Review by cgorantla — APPROVED** [review]

**Pull request merged**